[Samantha Low] iP

samlsm opened the pull request.

Commit:
Add Gradle support

samlsm committed:
Edited Duke.java

samlsm committed:
Edited test-ui files

samlsm committed:
Edited Duke.java files

samlsm committed:
Edited test-ui files

samlsm committed:
Edited Task class

samlsm committed:
Edited Duke.java

samlsm committed:
Edited test-ui files

samlsm committed:
Added different type of Task

samlsm committed:
Edited Duke

samlsm committed:
Edited all main files

samlsm committed:
Edited test-ui files for testing

samlsm committed:
Added delete function

samlsm committed:
Edited delete function

samlsm committed:
Added more test to test delete function

samlsm committed:
Added print template

samlsm committed:
Edited Duke to use template

samlsm committed:
Removed case conversion and added ENUM class

samlsm committed:
Edited error handling

samlsm committed:
Edited ui-test files

samlsm committed:
Add save function to Duke

Duke is unable to save task input and thus all tasks are lost once Duke terminates.

Let's create a hard disk and enable duke to write to that database

With tasks being save in hard disk whenever the task list changes, Duke can now save task inputs.

samlsm committed:
Edit text-ui-test

samlsm committed:
Add Date and Time functionality

Duke reads all date and time as String object.

Reading date and time as String limits the things Duke can do. Comparison of dates and times cannot be made as it is a String object.

Let's change the String object to LocalDate and LocalTime object instead.

In this way, Duke would be able to compare the Dates and organise tasks with respect to the dates on command.

samlsm committed:
Edit test-ui files to test date and time functionality

samlsm committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java
#	src/main/java/Event.java
#	text-ui-test/EXPECTED.TXT

samlsm committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Duke.java
#	src/main/java/Event.java
#	text-ui-test/EXPECTED.TXT

samlsm committed:
Edit text-ui-test to test merge of Level-7 and Level-8

samlsm committed:
Refactor Duke to use more OOP.

Duke class has too many functionalities/ purpose which violates the OOP principle.

Let's add in more classes and make use of inheritance to group classes with similar functionalities together.

Including more OOP allows the functionalities in Duke class to be passed to other classes. Adhere to OOP principle more adequately.

samlsm committed:
Refactor Duke to use more OOP.

Duke class has too many functionalities/ purpose which violates the OOP principle.

Let's add in more classes and make use of inheritance to group classes with similar functionalities together.

Including more OOP allows the functionalities in Duke class to be passed to other classes. Adhere to OOP principle more adequately.

samlsm committed:
Edit text-ui-test to test Duke

samlsm committed:
Edit text-ui-test to test Duke.

samlsm committed:
Organize into Packages.

Increase in number of classes makes it harder to search/locate class and to control access since all classes are not in any packages.

Let's organise the classes into multiple packages.

Makes it easier to deal with the increasing number of packages.

samlsm committed:
Address bug of scanner not closing after exitCommand.

Running with Duke with jave -jar command shows that scanner remains open and program does not terminate properly after exitCommand given.

Let's add a boolean to check for exitCommand and close scanner appropriately.

exitCommand now terminates program successfully

samlsm committed:
Override equals method of all Command package class.

The equals method of all Command classes compares identity rather than type of command.

Let's override the equals method so that it checks the type of command instead.

This is more logical as there is no difference between two commands of the same class. Mainly use for JUnit test.

samlsm committed:
Add ParserStub class for JUnit Test.

Parser class has dependencies on Task class.

Let's create a ParserStub class to remove this dependency.

This ensures that the methods in Parser class can be tested in isolation.

samlsm committed:
Add test classes to perform JUnit Tests.

samlsm committed:
Update file with javaDocs

samlsm committed:
Check coding standards

samlsm committed:
Refactor code to address merge conflict.

Merge conflict exists when trying to merge Master with branch-A-CodingStandard.

Let's address the merge conflict by editing Master.

Once all merge conflicts are addressed, branch can be merged with Master successfully.

samlsm committed:
Add find functionality.

Duke unable to print tasks by searching for a keyword.

Let's add a FindCommand to with find functionalities.

Users can now find a task by searching for a keyword.

samlsm committed:
Edited Duke files to deal with merge conflicts

samlsm committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/Storage.java
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/Ui.java

samlsm committed:
Refactor minor issues

samlsm committed:
Refactor minor issues

samlsm committed:
Refactor minor issues relating to find functionality

samlsm committed:
Update javaDocs

samlsm committed:
Edit find functionality to take in multiple keywords

samlsm committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

samlsm committed:
Use Checkstyle to detect coding style violations

Check for coding style violations previously but did not managed to identify all.

Let's make use of CheckStyle in gradle to check for coding style violations.

Ensures all files meets the checkStyle requirements

samlsm committed:
Add GUI client to Duke

samlsm committed:
Refactor Duke.

Duke's Storage has huge dependencies on TaskList.

Let's reduce this dependency as it complicates the code.

In this way, code is simplified making it easier to accommodate changes and debugging.

samlsm committed:
Add Main.java

samlsm committed:
Update files

samlsm committed:
Merge branch 'branch-Level-10'

samlsm committed:
Update javaDoc and checkstyle

samlsm committed:
Update relevant files

samlsm committed:
Add more test files

samlsm committed:
Update files

samlsm committed:
Update files for javaFXML

samlsm committed:
Update javaDocs

samlsm committed:
Move stubs to test directory

nicholasyeo commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,80 @@
+ * The <code>Parser</code> reads a command from user and perform various checks to determine the next action.
+ */
+public class Parser {
+    private static final String ignoreCase = "(?i)";
+    private static final String wildcard = "(.*)";
```
Perhaps these variable names could follow the format for naming constants?

samlsm replied:
Seems like I overlooked that... Thanks for the review!

nicholasyeo commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,26 @@
+ */
+public interface Command {
+
+    /**
+     * Executes this <code>Command</code>.
+     *
+     * @param command String representation of the command to be executed
+     * @param storage Storage of this <code>Duke</code>
+     * @param ui Ui containing all prints for user interactions
+     * @param taskList List of task for this <code>Duke</code>
+     * @return a string representation of the message informing user if the command has been successfully executed
+     * @throws DukeException if command fails to execute
+     */
+    String execute(String command, Storage storage, Ui ui, TaskList taskList) throws DukeException;
```
I noticed that similar header comments have been written for this method in the other classes that implement this interface. Not sure if its just me but, I feel that it suffices to write a general description of this method here and omit the comments in the other classes!

nicholasyeo reviewed (commented):
LGTM, almost. Some trivial coding standard violations such as naming constants. But, overall is good and variable names are generally clear and code is easy to read!

samlsm committed:
Update name of file

samlsm committed:
Add assertions to duke.

Did not make use of assertions to locate potential bugs.

Let's add assertions to document important assumptions that should hold at various points in the code.

By implementing assertions, location of potential bugs is now easier.

samlsm committed:
Refactor to improve code quality.

samlsm committed:
Add streams and lambdas to existing Duke.

samlsm committed:
Refactor to improve code quality.

samlsm committed:
Refactor to improve code quality.

samlsm committed:
Merge pull request #1 from samlsm/branch-A-Assertions

Branch a assertions

samlsm committed:
Merge remote-tracking branch 'origin/master'

samlsm committed:
Merge branch 'master' into branch-A-CodeQuality

samlsm committed:
Merge branch 'master' into branch-A-Streams

samlsm committed:
Merge pull request #2 from samlsm/branch-A-CodeQuality

Branch a code quality

samlsm committed:
Merge branch 'master' into branch-A-Streams

samlsm committed:
Merge pull request #3 from samlsm/branch-A-Streams

Branch a streams

samlsm committed:
Add undo functionality.

Unable to undo a specified command. Eg. Once a task is marked as done, cannot be reversed.

Let's add undo functionality to Duke so that user's have a way to undo an action.

In this way, user can revert back to a previous state if they make any mistake.

samlsm committed:
Add a Command that allows help.

There is no way of teaching user how to execute each Command and the correct format for it.

Let's add an additional help command which prints all the Commands' format.

In this way, user will know how to execute each Command and the type of Commands available.

samlsm committed:
Add UndoableCommand interface.

Command which cannot be undone have to implement the undo method.

Let's add in an interface that extends the Command interface.

This allows only Command which can be undone to implement the undo method. Command which cannot be undone would not need to know the existence of such undo method.

samlsm committed:
Add gradle.yml

samlsm committed:
Refactor to meet checkstyle

samlsm committed:
Edit LocalDate.

samlsm committed:
Checkstyle.

samlsm committed:
Edit text-ui-test.

samlsm committed:
Remove I/O redirection test.

samlsm committed:
Remove windows platform.

samlsm committed:
Improve the GUI.

samlsm committed:
Add more assertions.

samlsm committed:
Update README.md.

samlsm committed:
Edit format of printout.

samlsm committed:
Set theme jekyll-theme-cayman

samlsm committed:
Set theme jekyll-theme-modernist

samlsm committed:
Update README.md.

samlsm committed:
Merge remote-tracking branch 'origin/master'

samlsm committed:
Update README.md.

samlsm committed:
Remove storageTest.

samlsm committed:
Update README.md.

samlsm committed:
Update README.md.

samlsm committed:
Address bug related to Windows.

samlsm committed:
Update README.md.

samlsm committed:
Update README.md.

samlsm committed:
Update README.md.

samlsm committed:
Set theme jekyll-theme-hacker

samlsm committed:
Address time format issue.

samlsm committed:
Merge remote-tracking branch 'origin/master'

samlsm committed:
Edit help message.

samlsm committed:
Update Ui picture.

samlsm committed:
Update README.md
